## Guided session 1

This is one of the notebooks for the first session of the [Machine Learning workshop series at Harvey Mudd College](http://www.aashitak.com/ML-Workshops/). A lot of cells in the notebook are left blank for you to fill in.


Let us first import the python module `pandas`.

In [1]:
import pandas as pd

Pandas: an excellent tool to work with datasets

Dataframes: the central data structure of pandas library
- Evolved out of tables
- Most suitable for data manipulation tasks  

Pandas is built on top of numpy. The crucial difference between numpy matrices and pandas Dataframes is that the columns in a Dataframe can be of different datatypes such as numerical, categorical, textual, etc.

First, we load the Groundhog Day dataset [available here](https://www.kaggle.com/groundhogclub/groundhog-day/home)

In [2]:
# Uncomment the below two lines only if using Google Colab
# from google.colab import files
# uploaded = files.upload()
df = pd.read_csv('archive.csv')

The legend behind Groundhog Day tradition goes likes this:

> Thousands gather at Gobbler’s Knob in Punxsutawney, Pennsylvania, on the second day of February to await the spring forecast from a groundhog known as Punxsutawney Phil. According to legend, if Phil sees his shadow the United States is in store for six more weeks of winter weather. But, if Phil doesn’t see his shadow, the country should expect warmer temperatures and the arrival of an early spring.

[The dataset](https://www.kaggle.com/groundhogclub/groundhog-day/home) consists of the temperature records as well as the sightings of groundhog Phil.

In [4]:
# df

As it turns out to be rather big dataset to display, we can comment the above cell by adding `#` in front of `df` and run it again to get rid of the output. 

Next, let's check the numbers of rows and columns in the dataset.

In [22]:
df.shape

(132, 10)

So, the dataset consists of 132 rows and 10 columns. 

We use `head()` function to peek into the first 5 rows (or any number of rows by using `head(n)`). 

In [3]:
df.head()

Year Punxsutawney Phil  February Average Temperature  \
0  1886         No Record                           NaN   
1  1887       Full Shadow                           NaN   
2  1888       Full Shadow                           NaN   
3  1889         No Record                           NaN   
4  1890         No Shadow                           NaN   

   February Average Temperature (Northeast)  \
0                                       NaN   
1                                       NaN   
2                                       NaN   
3                                       NaN   
4                                       NaN   

   February Average Temperature (Midwest)  \
0                                     NaN   
1                                     NaN   
2                                     NaN   
3                                     NaN   
4                                     NaN   

   February Average Temperature (Pennsylvania)  March Average Temperature  \
0                                          NaN                        NaN   
1                                          NaN                        NaN   
2                                          NaN                        NaN   
3                                          NaN                        NaN   
4                                          NaN                        NaN   

   March Average Temperature (Northeast)  March Average Temperature (Midwest)  \
0                                    NaN                                  NaN   
1                                    NaN                                  NaN   
2                                    NaN                                  NaN   
3                                    NaN                                  NaN   
4                                    NaN                                  NaN   

   March Average Temperature (Pennsylvania)  
0                                       NaN  
1                                       NaN  
2                                       NaN  
3                                       NaN  
4                                       NaN

This has not been particularly useful since the records for the first few years are missing. Find the built-in function to display **the last 10 rows** from the DataFrame and use it below:

Year Punxsutawney Phil  February Average Temperature  \
122       2008       Full Shadow                         34.70   
123       2009       Full Shadow                         36.77   
124       2010       Full Shadow                         31.80   
125       2011         No Shadow                         33.04   
126       2012       Full Shadow                         37.51   
127       2013         No Shadow                         34.77   
128       2014       Full Shadow                         32.13   
129       2015       Full Shadow                         32.99   
130       2016         No Shadow                         39.47   
131  1901-2000               NaN                         33.82   

     February Average Temperature (Northeast)  \
122                                      25.4   
123                                      26.2   
124                                      25.9   
125                                      24.1   
126                                      30.6   
127                                      25.3   
128                                      21.0   
129                                      12.1   
130                                      28.1   
131                                      22.5   

     February Average Temperature (Midwest)  \
122                                    32.1   
123                                    35.9   
124                                    28.2   
125                                    34.4   
126                                    38.6   
127                                    33.4   
128                                    27.1   
129                                    23.4   
130                                    36.9   
131                                    32.9   

     February Average Temperature (Pennsylvania)  March Average Temperature  \
122                                         27.7                      41.86   
123                                         29.6                      42.87   
124                                         25.8                      43.57   
125                                         28.6                      43.07   
126                                         34.3                      50.41   
127                                         27.4                      40.91   
128                                         22.8                      40.51   
129                                         16.1                      45.39   
130                                         30.8                      47.50   
131                                         26.5                      41.50   

     March Average Temperature (Northeast)  \
122                                   31.6   
123                                   34.1   
124                                   39.6   
125                                   33.0   
126                                   43.4   
127                                   32.7   
128                                   26.0   
129                                   27.8   
130                                   38.8   
131                                   32.3   

     March Average Temperature (Midwest)  \
122                                 42.0   
123                                 45.7   
124                                 44.6   
125                                 44.2   
126                                 56.3   
127                                 37.3   
128                                 38.0   
129                                 41.6   
130                                 49.5   
131                                 42.4   

     March Average Temperature (Pennsylvania)  
122                                      35.6  
123                                      38.2  
124                                      42.0  
125                                      36.3  
126                                      47.7  
127                                      33.9  
128                        

Let's see the columns in the DataFrame. 

In [10]:
df.columns

Index(['Year', 'Punxsutawney Phil', 'February Average Temperature',
       'February Average Temperature (Northeast)',
       'February Average Temperature (Midwest)',
       'February Average Temperature (Pennsylvania)',
       'March Average Temperature', 'March Average Temperature (Northeast)',
       'March Average Temperature (Midwest)',
       'March Average Temperature (Pennsylvania)'],
      dtype='object')

### Selecting rows and columns from the dataframe

Let's say we want to see only the columns related to the groundhog Phil sighting as well as the average temperatures in Febrauary in Pennsylvania and get rid of the rest of the columns. This can be accomplished using double brackets:

In [11]:
df[['Punxsutawney Phil', 'February Average Temperature (Pennsylvania)']].head()

Punxsutawney Phil  February Average Temperature (Pennsylvania)
0         No Record                                          NaN
1       Full Shadow                                          NaN
2       Full Shadow                                          NaN
3         No Record                                          NaN
4         No Shadow                                          NaN

Since we do not want all the rows in the output, we have used `head()` function at the end. 

We can also select rows based on conditions. Let's say, we want to observe only those years when there is no record for groundhog. Hint: Use `df['Punxsutawney Phil'] == "No Record"` in the conditional.

In [12]:
df[df["Punxsutawney Phil"]=="No Record"]

Year Punxsutawney Phil  February Average Temperature  \
0   1886         No Record                           NaN   
3   1889         No Record                           NaN   
5   1891         No Record                           NaN   
6   1892         No Record                           NaN   
7   1893         No Record                           NaN   
8   1894         No Record                           NaN   
9   1895         No Record                         26.60   
10  1896         No Record                         35.04   
11  1897         No Record                         33.39   
13  1899         No Record                         25.50   
16  1902         No Record                         31.46   
57  1943         No Record                         36.28   

    February Average Temperature (Northeast)  \
0                                        NaN   
3                                        NaN   
5                                        NaN   
6                                        NaN   
7                                        NaN   
8                                        NaN   
9                                       15.6   
10                                      22.2   
11                                      23.6   
13                                      18.1   
16                                      20.1   
57                                      23.9   

    February Average Temperature (Midwest)  \
0                                      NaN   
3                                      NaN   
5                                      NaN   
6                                      NaN   
7                                      NaN   
8                                      NaN   
9                                     21.9   
10                                    33.5   
11                                    34.7   
13                                    22.2   
16                                    23.6   
57                                    35.4   

    February Average Temperature (Pennsylvania)  March Average Temperature  \
0                                           NaN                        NaN   
3                                           NaN                        NaN   
5                                           NaN                        NaN   
6                                           NaN                        NaN   
7                                           NaN                        NaN   
8                                           NaN                        NaN   
9                                          17.0                      39.97   
10                                         26.6                      38.03   
11                                         27.9                      38.79   
13                                         20.0                      37.63   
16                                         21.0                      41.58   
57                                         27.6                      38.50   

    March Average Temperature (Northeast)  \
0                                     NaN   
3                                     NaN   
5                                     NaN   
6                                     NaN   
7                                     NaN   
8                                     NaN   
9                                    27.6   
10                                   25.3   
11                                   32.0   
13                                   29.3   
16                                   37.1   
57                                   30.5   

    March Average Temperature (Midwest)  \
0                                   NaN   
3                                   NaN   
5                                   NaN   
6                                   NaN   
7                                   NaN   
8                                   NaN   
9                                  40.2   
10                                 36.9   
11         

Now, we want to restrict the above output further *to exclude the entries before the year 1895*.   
Hint: Add another conditional to the above code using `&` and make sure to wrap the two conditionals with parenthesis.   

Year Punxsutawney Phil  February Average Temperature  \
10  1896         No Record                         35.04   
11  1897         No Record                         33.39   
13  1899         No Record                         25.50   
16  1902         No Record                         31.46   
57  1943         No Record                         36.28   

    February Average Temperature (Northeast)  \
10                                      22.2   
11                                      23.6   
13                                      18.1   
16                                      20.1   
57                                      23.9   

    February Average Temperature (Midwest)  \
10                                    33.5   
11                                    34.7   
13                                    22.2   
16                                    23.6   
57                                    35.4   

    February Average Temperature (Pennsylvania)  March Average Temperature  \
10                                         26.6                      38.03   
11                                         27.9                      38.79   
13                                         20.0                      37.63   
16                                         21.0                      41.58   
57                                         27.6                      38.50   

    March Average Temperature (Northeast)  \
10                                   25.3   
11                                   32.0   
13                                   29.3   
16                                   37.1   
57                                   30.5   

    March Average Temperature (Midwest)  \
10                                 36.9   
11                                 44.0   
13                                 38.4   
16                                 43.9   
57                                 38.5   

    March Average Temperature (Pennsylvania)  
10                                      27.8  
11                                      36.9  
13                                      34.0  
16                                      38.8  
57                                      34.7

We can also get the number of years with no records using the `shape` attribute which gives us both the number of columns and the number of rows. Write the code to count the number of years that the groundhog saw his full shadow.

12

Write the code to count the number of years in which the groundhog Phil saw its full shadow and the February Average Temperature in Pennsylvania was less than 26.5.

45

Write the code to count the number of years in which the groundhog Phil saw its full shadow and the February Average Temperature in Pennsylvania was less than February Average Temperature in that year.

98

Is it common for Pennsylvania to have temperatures colder than average? Write down the code for counting the years when February Average Temperature in Pennsylvania is less than February Average Temperature in general.  
Hint: This requires only one condition and that can be achieved from removing one of the condition from the last code cell.

121

Please feel free to explore the data on your own later. You can also design a hypothesis and test it using `scipy.stats`. 

### The `loc` and `iloc` methods

So far, we have seen how to retrieve either some select columns or certain rows based on conditionals. What if we want to slice off a portion of the dataframe with some specific rows and columns? We use `.loc[]` or `.iloc[]` methods for this purpose. 
* `.iloc[]` method is primarily integer position based and gets rows/columns at particular positions in the index (so it only takes integers). 
* `loc[]` method is label based and gets rows/columns with particular labels from the index. 

In [34]:
df.loc[:3, ['Year', 'Punxsutawney Phil']]

Year Punxsutawney Phil
0  1886         No Record
1  1887       Full Shadow
2  1888       Full Shadow
3  1889         No Record

In [36]:
# df.iloc[:3, ['Year', 'Punxsutawney Phil']] # This will give an error
# Comment the above line of code and uncomment the below one
df.iloc[:3, [0, 1]] 

Year Punxsutawney Phil
0  1886         No Record
1  1887       Full Shadow
2  1888       Full Shadow

 Just like Python lists, the indexing in pandas start from 0. To get the **11th row**, we use `df.iloc[10]` as below.

In [37]:
df.iloc[10]

Year                                                1896
Punxsutawney Phil                              No Record
February Average Temperature                       35.04
February Average Temperature (Northeast)            22.2
February Average Temperature (Midwest)              33.5
February Average Temperature (Pennsylvania)         26.6
March Average Temperature                          38.03
March Average Temperature (Northeast)               25.3
March Average Temperature (Midwest)                 36.9
March Average Temperature (Pennsylvania)            27.8
Name: 10, dtype: object

The `.iloc[]` method can slice any collection of rows, not necessarily consecutive. For example, we can slice data corresponding to every tenth year.

In [38]:
df.iloc[0:len(df):10]

Year Punxsutawney Phil  February Average Temperature  \
0    1886         No Record                           NaN   
10   1896         No Record                         35.04   
20   1906       Full Shadow                         33.69   
30   1916       Full Shadow                         33.03   
40   1926       Full Shadow                         37.67   
50   1936       Full Shadow                         25.23   
60   1946       Full Shadow                         34.88   
70   1956       Full Shadow                         32.27   
80   1966       Full Shadow                         31.24   
90   1976       Full Shadow                         39.56   
100  1986         No Shadow                         35.85   
110  1996       Full Shadow                         35.69   
120  2006       Full Shadow                         34.83   
130  2016         No Shadow                         39.47   

     February Average Temperature (Northeast)  \
0                                         NaN   
10                                       22.2   
20                                       22.0   
30                                       18.9   
40                                       20.5   
50                                       15.6   
60                                       21.4   
70                                       25.6   
80                                       23.2   
90                                       28.8   
100                                      22.3   
110                                      23.6   
120                                      26.2   
130                                      28.1   

     February Average Temperature (Midwest)  \
0                                       NaN   
10                                     33.5   
20                                     30.7   
30                                     30.1   
40                                     36.3   
50                                     24.0   
60                                     36.5   
70                                     36.1   
80                                     31.9   
90                                     41.1   
100                                    34.3   
110                                    33.4   
120                                    33.7   
130                                    36.9   

     February Average Temperature (Pennsylvania)  March Average Temperature  \
0                                            NaN                        NaN   
10                                          26.6                      38.03   
20                                          25.3                      36.10   
30                                          22.8                      42.15   
40                                          25.8                      39.65   
50                                          18.3                      43.43   
60                                          27.4                      46.78   
70                                          29.8                      40.75   
80                                          26.5                      43.00   
90                                          33.5                      42.31   
100                                         26.7                      46.18   
110                                         27.0                      39.02   
120                                         29.8                      42.62   
130                                         30.8                      47.50   

     March Average Temperature (Northeast)  \
0                                      NaN   
10                                    25.3   
20                                    25.7   
30                                    24.2   
40                                    25.9   
50                                    37.4   
60                                    41.6   
70                                    27.9   
80                                    34.1   
9

For convenience, we want to work with a subset of dataset, say the most recent 20 years of data. Complete the code below using `.iloc[]` method to give the last 20 entries of the dataset. 

Year Punxsutawney Phil  February Average Temperature  \
112       1998       Full Shadow                         38.71   
113       1999         No Shadow                         39.49   
114       2000       Full Shadow                         39.78   
115       2001       Full Shadow                         33.98   
116       2002       Full Shadow                         36.39   
117       2003       Full Shadow                         32.79   
118       2004       Full Shadow                         33.57   
119       2005       Full Shadow                         37.94   
120       2006       Full Shadow                         34.83   
121       2007         No Shadow                         32.41   
122       2008       Full Shadow                         34.70   
123       2009       Full Shadow                         36.77   
124       2010       Full Shadow                         31.80   
125       2011         No Shadow                         33.04   
126       2012       Full Shadow                         37.51   
127       2013         No Shadow                         34.77   
128       2014       Full Shadow                         32.13   
129       2015       Full Shadow                         32.99   
130       2016         No Shadow                         39.47   
131  1901-2000               NaN                         33.82   

     February Average Temperature (Northeast)  \
112                                      31.6   
113                                      28.2   
114                                      27.2   
115                                      26.0   
116                                      29.7   
117                                      20.0   
118                                      24.0   
119                                      26.2   
120                                      26.2   
121                                      18.0   
122                                      25.4   
123                                      26.2   
124                                      25.9   
125                                      24.1   
126                                      30.6   
127                                      25.3   
128                                      21.0   
129                                      12.1   
130                                      28.1   
131                                      22.5   

     February Average Temperature (Midwest)  \
112                                    40.9   
113                                    39.5   
114                                    40.1   
115                                    36.1   
116                                    36.8   
117                                    30.2   
118                                    33.1   
119                                    38.1   
120                                    33.7   
121                                    26.1   
122                                    32.1   
123                                    35.9   
124                                    28.2   
125                                    34.4   
126                                    38.6   
127                                    33.4   
128                                    27.1   
129                                    23.4   
130                                    36.9   
131                                    32.9   

     February Average Temperature (Pennsylvania)  March Average Temperature  \
112                                         35.8                      41.34   
113                                         31.7                      42.80   
114                                         31.6                      46.04   
115                                         31.5                      41.49   
116                                         33.9                      39.54   
117                                         23.8                      43.30   
118                                         27.0    

Use the `.loc[]` method to find out whether groundhog Phil casted full shadow or not in the year 2000. Hint: Use conditional for the row index.

114    Full Shadow
Name: Punxsutawney Phil, dtype: object

Could you use `iloc` with the conditional above? Check it out!

### Next step:
Please continue on to the [hands-on exercise](https://github.com/AashitaK/ML-Workshops/blob/master/Session%201/Exercise%201.ipynb). You are encouraged to work in groups of 2-4 for the exercise session. Please ask for help from the instructor and TAs.

### Acknowledgment:
* [Groundhog Day Forecasts and Temperatures](https://www.kaggle.com/groundhogclub/groundhog-day/home) dataset openly available in Kaggle is used for illustration.

**Note:**
The solutions for this exercise can be found [here](https://github.com/AashitaK/ML-Workshops/blob/master/Session%201/Guided%20session%201%20with%20solutions.ipynb).
